services/job_runner.go

```diff
@@ -162,218 +161,84 @@ func (rm *jobRunner) workerCount() int {
-		if err != nil {
-			return jr, wrapExecuteRunAtBlockError(jr, err)
-		}
-
-		lastRun := markCompletedIfRunnable(startTask(jr, nextTaskRun, prevResult, bn, store))
-		jr.TaskRuns[i+offset] = lastRun
-		logTaskResult(lastRun, nextTaskRun, i)
-		prevResult = lastRun.Result
+	logger.Infow(fmt.Sprintf("Processing task %s", currentTaskRun.Task.Type), []interface{}{"task", currentTaskRun.ID}...)
 
-		if err := store.Save(&jr); err != nil {
-			return jr, wrapExecuteRunAtBlockError(jr, err)
-		}
-		if !lastRun.Status.Runnable() {
-			break
-		}
-	}
+	result := adapter.Perform(input, store)
 
-	jr = jr.ApplyResult(prevResult)
-	logger.Infow("Finished current job run execution", jr.ForLogger()...)
-	return jr, wrapExecuteRunAtBlockError(jr, store.Save(&jr))
-}
+	logger.Infow(fmt.Sprintf("Finished processing task %s", currentTaskRun.Task.Type), []interface{}{
+		"task", currentTaskRun.ID,
+		"result", result.Status,
+	}...)
 
-func prepareJobRun(
-	jr models.JobRun,
-	store *store.Store,
-	bn *models.IndexableBlockNumber,
-) (models.JobRun, error) {
-	if jr.Status.CanStart() {
-		jr.Status = models.RunStatusInProgress
-	} else {
-		return jr, fmt.Errorf("Unable to start with status %v", jr.Status)
-	}
-	if err := store.Save(&jr); err != nil {
-		return jr, err
-	}
-	if jr.Result.HasError() {
-		return jr, jr.Result
-	}
-	return store.SaveCreationHeight(jr, bn)
+	return result
 }
 
-func logTaskResult(lr models.TaskRun, tr models.TaskRun, i int) {
-	logger.Debugw("Produced task run", "taskRun", lr)
-	logger.Debugw(fmt.Sprintf("Next task run %v %v", tr.Task.Type, tr.Result.Status), tr.ForLogger("taskidx", i, "result", lr.Result)...)
-}
+func executeRun(run *models.JobRun, store *store.Store) error {
+	logger.Infow("Processing run", run.ForLogger()...)
 
-func markCompletedIfRunnable(tr models.TaskRun) models.TaskRun {
-	if tr.Status.Runnable() {
-		return tr.MarkCompleted()
+	if !run.Status.Runnable() {
+		return fmt.Errorf("Run triggered in non runnable state %s", run.Status)
 	}
-	return tr
-}
 
-func startTask(
-	jr models.JobRun,
-	tr models.TaskRun,
-	input models.RunResult,
-	bn *models.IndexableBlockNumber,
-	store *store.Store,
-) models.TaskRun {
-	adapter, err := adapters.For(tr.Task, store)
-	if err != nil {
-		return tr.ApplyResult(tr.Result.WithError(err))
+	if !run.TasksRemain() {
+		return errors.New("Run triggered with no remaining tasks")
 	}
 
-	minConfs := utils.MaxUint64(
-		store.Config.MinIncomingConfirmations,
-		tr.Task.Confirmations,
-		adapter.MinConfs())
+	prevTaskRun, currentTaskRun := run.CurrentTaskRuns()
+	result := executeTask(run, prevTaskRun, currentTaskRun, store)
 
-	if !jr.Runnable(bn, minConfs) {
-		tr = tr.MarkPendingConfirmations()
-		tr.Result.Data = input.Data
-		return tr
+	currentTaskRun.ApplyResult(result)
+	if currentTaskRun.Status.PendingSleep() {
+		logger.Debugw("Task is sleeping", []interface{}{"run", run.ID}...)
+		if err := ResumeSleepingTask(run, store); err != nil {
```

Comment:
It was slightly awkward for me to read `PendingSleep` immediately after executing task to see it then `ResumeSleepingTask`. Conceptually I would think that would happen inside the `executeTask`, but it seems to happen later on inside this job runner since you've moved sleeping outside of the adapter.

Reply:
Yeah, maybe `Resume` is not the best name here.